domain/application/state/storage.go

```diff
@@ -78,6 +98,255 @@ VALUES ($storageToAdd.*)`, storageToAdd{})
 	return nil
 }
 
+type storageTemplate struct {
+	meta   charmStorage
+	params application.ApplicationStorageArg
+}
+
+// insertUnitStorage inserts the storage records need to record the intent
+// for the specified new unit and storage args. Records include:
+// - storage instance
+// - filesystem
+// - volume
+// - related attachment records
+// TODO(storage) - support attaching existing storage when adding a unit
+func (st *State) insertUnitStorage(
+	ctx context.Context, tx *sqlair.TX, appUUID coreapplication.ID, unitUUID coreunit.UUID, netNodeUUID string,
+	args []application.ApplicationStorageArg, poolKinds map[string]storage.StorageKind,
+) error {
+
+	// Reduce the count of new storage created for each existing storage
+	// being attached.
+	// TODO(storage) - implement this when unit machine storage can be supported
+	// (includes validateCharmStorageCountChange below)
+
+	templates := make([]storageTemplate, 0, len(args))
+	for _, arg := range args {
+		storageMeta, err := st.getApplicationCharmStorageByName(ctx, tx, appUUID, arg.Name)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Errorf(
+					"charm for application %q has no storage called %q",
+					appUUID, arg.Name,
+				).Add(applicationerrors.StorageNameNotSupported)
+			}
+			return errors.Errorf("getting charm storage metadata for storage name %q application %q: %w", arg.Name, appUUID, err)
+		}
+
+		if arg.Count == 0 {
+			continue
+		}
+		templates = append(templates, storageTemplate{
+			meta:   storageMeta,
+			params: arg,
+		})
+	}
+
+	app := applicationID{ID: appUUID}
+	selectCharmStmt, err := st.Prepare(
+		`SELECT &applicationCharmUUID.charm_uuid FROM application WHERE uuid = $applicationID.uuid`,
+		app, applicationCharmUUID{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	for _, t := range templates {
+		if err := validateCharmStorageCountChange(t.meta, 0, t.params.Count); err != nil {
+			return err
+		}
```

Comment:
This should be done in service. Charm metadata is immutable.

Reply:
Charm metadata might be immutable but the charm used by a unit isn't.